_posts/2025-02-28-lattice-bitset.md

```diff
@@ -0,0 +1,801 @@
+<polygon fill="black" stroke="black" points="168.56,-140.77 178.63,-144.04 172.79,-135.2 168.56,-140.77"/>
+</g>
+<!-- any -->
+<g id="node8" class="node">
+<title>any</title>
+<ellipse fill="none" stroke="black" cx="200.92" cy="-234" rx="74.33" ry="18"/>
+<text text-anchor="middle" x="200.92" y="-229.8" font-family="Times,serif" font-size="14.00">Object/Top/Any</text>
+</g>
+<!-- int_str&#45;&gt;any -->
+<g id="edge10" class="edge">
+<title>int_str-&gt;any</title>
+<path fill="none" stroke="black" d="M93.99,-178.36C113.35,-188.25 138.76,-201.24 159.96,-212.07"/>
+<polygon fill="black" stroke="black" points="158.13,-215.07 168.63,-216.5 161.32,-208.83 158.13,-215.07"/>
+</g>
+<!-- int_list&#45;&gt;any -->
+<g id="edge11" class="edge">
+<title>int_list-&gt;any</title>
+<path fill="none" stroke="black" d="M200.92,-180.3C200.92,-187.59 200.92,-196.27 200.92,-204.46"/>
+<polygon fill="black" stroke="black" points="197.42,-204.38 200.92,-214.38 204.42,-204.38 197.42,-204.38"/>
+</g>
+<!-- str_list&#45;&gt;any -->
+<g id="edge12" class="edge">
+<title>str_list-&gt;any</title>
+<path fill="none" stroke="black" d="M299.41,-177.85C281.96,-187.62 258.83,-200.57 239.36,-211.47"/>
+<polygon fill="black" stroke="black" points="237.91,-208.28 230.89,-216.22 241.33,-214.39 237.91,-208.28"/>
+</g>
+</g>
+</svg>
+</figure>
+
+See how we have two new layers! We have a Bottom (empty) type now. For example,
+`join(Bottom, Bottom)` is `Bottom` (it remains empty), and `join(Bottom, List)`
+is `List` (no new elements).
+
+We can also address the set of all ints and all strings with `Int|String`,
+which is ordered above each of the invidual int and string sets.
+
+As I alluded to earlier, we can conveniently represent a bottom element in our
+bitset type:
+
+```c++
+enum Type {
+    Bottom = 0,       // 0b000
+    Int    = 1 << 0,  // 0b001
+    List   = 1 << 1,  // 0b010
+    String = 1 << 2,  // 0b100
+    Top    = 7,       // 0b111
+};
+
+
+enum Type join(enum Type left, enum Type right) {
+    return left | right;
+}
+
+// Ask if `left` is a subtype of `right`
+bool is_subtype(enum Type left, enum Type right) {
+    return (left & right) == left;
+}
+```
+
+<!-- TODO(max): Maybe talk about intersection from `if isinstance` refinement,
+GuardType refinement, or could_be -->
+
+This type representation is neat, but we can go further. Sometimes, you know
+more than just the *type* of an object: you know exactly what object it is.
+We'll introduce what Cinder calls *specialization*.
+
+## Specialization
+
+In addition to our existing type lattice, Cinder keeps a second lattice called
+a specialization. Its job is to keep track of what specific object an SSA value
+is. For example, if we know that `a = 5`, its type would be `Int` and its
+specialization would be `5`.
+
+The specialization lattice is much simpler:
+
+```c
+struct Spec {
+    enum {
+        SpecTop,
+        SpecInt,
+        SpecBottom,
+    } spec_kind;
+    int value;
+};
+```
```

Comment:
Any chance this can be written as a tagged enum using pseudocode? I get wanting it to be C, but it would be easier to grasp at a glance if it were

```
enum {
    SpecTop,
    SpecInt(value),
    SpecBottom,
} spec_kind;
```